### ISSUE TITLE
Sourcemap upload freeze when silent: true option is set

### ISSUE THREAD

**Issue by konclave** [getsentry/sentry-webpack-plugin#187]
Sometimes it logs "> Bundling files for upload... filename", sometimes nothing, sometimes everything goes well. But if no "silent: true" is set, everything is okay on every build

**Comment by kamilogorek** [comment]
Unfortunately, there's not much I can say here without any reproducible use-case or execution logs. I'm just not able to tell where to even start debugging this.

**Comment by konclave** [comment]
How do I provide more info about that? Seems like setting debug option to true doesn't help to get any additional information

**Comment by kamilogorek** [comment]
You can try to use `SENTRY_LOG_LEVEL=debug` env flag to make `sentry-cli` itself produce the logs, maybe this is the place where it gets stuck somethow.

**Comment by konclave** [comment]
It helped. I've got a long log of source files and source maps list, requests with 200 and 208 response statuses and it ends with 'skipping update nagger because session is not attended'. After that my build pipeline stuck.

**Comment by kamilogorek** [comment]
Can you provide that log somewhere?

**Comment by konclave** [comment]
https://drive.google.com/file/d/1JZTDrYrlBmgjzbMynmy6PTxxdAQ9I1HS/view?usp=sharing

**Comment by kamilogorek** [comment]
@konclave nothing that'd indicate the issue, unfortunately. We may also try to add `debug: true` to webpack plugin config. This will show whether we're stuck in the plugin itself or webpack is not handing the exit code correctly.

**Comment by konclave** [comment]
Setting `debug: true` options brings nothing new into the log. And it looks like setting `dryRun: true` option lead to a successful build. I've tried to restart a build something about 10 times and it didn't stuck.

**Comment by kamilogorek** [comment]
This means that there's most likely connection issue between your env and Sentry. We may verify this in the following way. In your logs, there's like starting with `sentry-cli was invoked with the following command line`.
Copy this whole command (and strip all the quotes, it's logging thing) and invoke it first "as is", and then try with `--wait` flag to see if files are processed correctly. Eg.

```
sentry-cli releases files foobar upload-sourcemaps dist --ignore node_modules
//
sentry-cli releases files foobar upload-sourcemaps dist --ignore node_modules --wait
```

I also wonder why does your log contain whole execution flow twice. It looks like you'd include SentryWebpackPlugin twice somehow. Maybe it's somehow merging dev/prod/common configs together?

**Comment by konclave** [comment]
Should I replace webpack plugin with those commands? All those things are happening on our CI server. So it's pretty hard to execute a separate command there. 
Should I call those commands one after another in one CI cycle? It will upload the same source maps twice.
Also there's no `silent` key for sentry-cli, and the issue happens only when `silent: true` param is set for webpack plugin.
I didn't found any duplication in a log. Maybe I've just pasted it twice, sorry.

**Comment by shulamit74** [comment]
+1 on the issue.
When buildijg with `silent: true` the build is just stuck. 
When building with `silent:false` the build succeeds and I get this error for a file in the repository:
```
components/DocumentHeader/MyLogo.js (no sourcemap ref)
- warning: could not determine a source map reference (Could not auto-detect referenced sourcemap for components/DocumentHeader/MyLogo.js.)

```

**Comment by shulamit74** [comment]
Update: removing the problematic file, it's still stuck with `silent: true`. I can provide more logs if needed.

**Comment by haoxl3** [comment]
@shulamit74 I have the same as problem with you, How did you solve it , please ?

**Comment by twelve17** [comment]
I'm running into the same issue (with `silent: true` the build hangs).  I've ran the `webpack` command with `SENTRY_LOG_LEVEL=debug `.  It seems to hang after this:

```
  DEBUG   2021-04-19 22:16:46.023318520 +00:00 > POST /api/0/organizations/<our org>/releases/<version>/assemble/ HTTP/1.1
  DEBUG   2021-04-19 22:16:46.023350935 +00:00 > Host: sentry.io
  DEBUG   2021-04-19 22:16:46.023370415 +00:00 > Accept: */*
  DEBUG   2021-04-19 22:16:46.023408395 +00:00 > Connection: TE
  DEBUG   2021-04-19 22:16:46.023433649 +00:00 > TE: gzip
  DEBUG   2021-04-19 22:16:46.023453140 +00:00 > User-Agent: sentry-cli/1.64.0
  DEBUG   2021-04-19 22:16:46.023491741 +00:00 > Authorization: ......
  DEBUG   2021-04-19 22:16:46.023511043 +00:00 > Content-Type: application/json
  DEBUG   2021-04-19 22:16:46.023529730 +00:00 > Content-Length: 109
  ```
  
  response
  
  ```
  DEBUG   2021-04-19 22:16:46.170485539 +00:00 < HTTP/1.1 200 OK
  DEBUG   2021-04-19 22:16:46.170535478 +00:00 < Server: nginx
  DEBUG   2021-04-19 22:16:46.170601188 +00:00 < Date: Mon, 19 Apr 2021 22:16:46 GMT
  DEBUG   2021-04-19 22:16:46.170625934 +00:00 < Content-Type: application/json
  DEBUG   2021-04-19 22:16:46.170655837 +00:00 < Content-Length: 38
  DEBUG   2021-04-19 22:16:46.170677081 +00:00 < Connection: keep-alive
  DEBUG   2021-04-19 22:16:46.170915415 +00:00 < allow: POST, OPTIONS
  DEBUG   2021-04-19 22:16:46.170961325 +00:00 < access-control-allow-methods: POST, OPTIONS
  DEBUG   2021-04-19 22:16:46.170995582 +00:00 < access-control-allow-headers: X-Sentry-Auth, X-Requested-With, Origin, Accept, Content-Type, Authentication, Authorization, Content-Encoding
  DEBUG   2021-04-19 22:16:46.171018808 +00:00 < access-control-expose-headers: X-Sentry-Error, Retry-After
  DEBUG   2021-04-19 22:16:46.171053529 +00:00 < access-control-allow-origin: *
  DEBUG   2021-04-19 22:16:46.171226573 +00:00 < vary: Accept-Language, Cookie
  DEBUG   2021-04-19 22:16:46.171275722 +00:00 < content-language: en
  DEBUG   2021-04-19 22:16:46.171297989 +00:00 < x-frame-options: deny
  DEBUG   2021-04-19 22:16:46.171459597 +00:00 < x-content-type-options: nosniff
  DEBUG   2021-04-19 22:16:46.171481914 +00:00 < x-xss-protection: 1; mode=block
  DEBUG   2021-04-19 22:16:46.171530467 +00:00 < x-envoy-attempt-count: 1
  DEBUG   2021-04-19 22:16:46.171559520 +00:00 < x-envoy-upstream-service-time: 29
  DEBUG   2021-04-19 22:16:46.171719002 +00:00 < x-served-by: getsentry-web-default-production-79d578896c-8w9cm
  DEBUG   2021-04-19 22:16:46.171773172 +00:00 < x-served-by: lb-48
  DEBUG   2021-04-19 22:16:46.171933154 +00:00 < Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
  DEBUG   2021-04-19 22:16:46.172092551 +00:00 response status: 200
  DEBUG   2021-04-19 22:16:46.172178218 +00:00 body: {"state":"created","missingChunks":[]}
  ```

**Comment by twelve17** [comment]
I've ran the sentry-cli (copied from my webpack plugin debug output) and it never hangs, regardless of the presence of the `--wait` flag. So my next suspicion is that there is something different when ran in an interactive shell vs. not. 

I see in the [centry cli source](https://github.com/getsentry/sentry-cli/blob/e0cfb090475f21405c230f3b9798da85d3a325dd/js/helper.js#L166) that the `execute` helper changes the value of `stdio` to `pipe`  when `slient` is true.  Upon some googlling, I found instances of pipes causing `spawn` to hang if a pipe is waiting for data that it never receives.

When the script is running and it hangs, on a separate shell, I run `strace` on the `sentry-cli` process, and I see it is just hanging on a `wait`:

```
strace: Process 413 attached
futex(0x1641ca8, FUTEX_WAIT_PRIVATE, 4294967295, NULL^Cstrace: Process 413 detached
```

I'm going to keep diving into this theory to see if I find anything.

**Comment by twelve17** [comment]
Per the [node stdio docs](https://nodejs.org/api/child_process.html#child_process_options_stdio):

> 'ignore': Instructs Node.js to ignore the fd in the child. While Node.js will always open fds 0, 1, and 2 for the processes it spawns, setting the fd to 'ignore' will cause Node.js to open /dev/null and attach it to the child's fd.

This kind of sounds (to me) what the intention behind `silent` is, though I could be missing some other functionality that requires use of `pipe`.  Either way,  as a preliminary test, I changed the `stdio` STDOUT option from `pipe` to `ignore`.  I have been running the build several times, and it has not yet hung.  So this is a promising lead so far.

**Comment by twelve17** [comment]
Ok, I think what might be the problem is that the sentry CLI child process created by `execute` is sending data to its `stdout`, and the parent is not consuming the stream. Eventually, the stream fills up and the process will hang (see https://github.com/nodejs/node-v0.x-archive/issues/6764 for an example). 

To test this, I edited the `execute` function to add a event handler to consume `stdout` from the child:

```
      pid.stdout.on('data', data => {
        console.log("Sentry CLI helper: STDOUT DATA:", data);
      });
```

When I run the build, despite having `silent: true`,  there is data being sent to `stdout` by the child process:

```
Sentry CLI helper: STDOUT DATA: > Found 668 release files

Sentry CLI helper: STDOUT DATA: > Analyzing 668 sources

Sentry CLI helper: STDOUT DATA: > Rewriting sources

Sentry CLI helper: STDOUT DATA: > Adding source map references

Sentry CLI helper: STDOUT DATA: > Bundled 668 files for upload

⠁ 
Sentry CLI helper: STDOUT DATA: > Uploaded release files to Sentry
Sentry CLI helper: STDOUT DATA: > File upload complete (processing pending on server)
```

Notably, the above change consumes the `stdout` stream, so the build does not hang.

**Comment by kamilogorek** [comment]
Great investigation, thanks @twelve17 
Fixed it here https://github.com/getsentry/sentry-cli/pull/949 and will update `sentry-cli` dependency once its merged.

**Comment by twelve17** [comment]
@kamilogorek great to hear, and thanks for the quick response!

**Comment by kamilogorek** [comment]
This should be fixed in `1.15.1`. Cheers!
